rfd/0-job-worker.md

```diff
@@ -0,0 +1,247 @@
+
+The server will initially reexecute itself, using `exec.Command()`, as follows:
+
+```sh
+/proc/self/exe child [serve flags] -- command [args]...
+```
+
+When calling this, it will set the command's `SysProcAttr` with:
+
+```go
+syscall.SysProcAttr{
+    Cloneflags: syscall.CLONE_NEWPID | // New pid namespace
+        syscall.CLONE_NEWNS | // New mount namespace group
+        syscall.CLONE_NEWNET, // New network namespace
+    Unshareflags: syscall.CLONE_NEWNS, // Isolate process mounts from host
+}
+```
+
+In order to provide proper accounting of the job the server will internally track of the execution with:
+
+- buffers for `stdout` and `stderr`
+- process status (e.g. running, complete)
+- exit code
+
+Once reexecuted, the `child` command has several jobs to do. It has to remount the `/proc` filesystem, create and configure the cgroup, set the cgroup for the pid and fully replace the execution, using `syscall.Exec()`, with the job binary.
+
+##### Job Status
+
+Job status is extremely simple and only returns a status code and an optional exit_code. The status code is one of:
+
+- running: the job has been started and has not yet completed
+- complete: the job completed successfully on its own
+- stopped: the job was manually stopped, this takes precedence over complete
+
+##### Job Output Storage
+
+Output for each job will be maintained, in memory, as long as the worker implementation exists. This poses a potential memory leak, but for the purposes of this challenge, is an acceptable solution. Later designs would likely need to write job output to disk or possibly even remote storage.
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
+
+Internally, the library will maintain a buffer containing the job output. The job itself will be configured to use this buffer for `stdout` and `stderr` and the `Write()` method will be goroutine safe. When a client calls `JobOutput()` the library will create a new `io.ReadCloser` that independently, and with goroutine safety, reads through to the end of the output buffer. If the job has already completed at this time, `io.EOF` will be returned. If not, subsequent calls to `Read()` will block until there is either more output to return or the job ends in which case `io.EOF` is returned. When the job calls `Write()` it will signal connected readers that new data has been written to the output buffer. The readers will in turn be able to unblock and return their `Read()` calls to connected clients at that time. If the user closes the `io.ReadCloser`, any pending `Read()` operations will unblock with `io.EOF` and any goroutines and other associated resources with the reader will be released.
```

Comment:
> If the user closes the `io.ReadCloser`, any pending `Read()` operations will unblock with `io.EOF` and any goroutines and other associated resources with the reader will be released.

nit: I would probably do an ErrClosed (or something like that) so callers can distinguish closed from a "true" EOF.

Reply:
I thought about that too. I even looked to see if there was a preexisting idiomatic error code for this purpose (none that I could find). My goal is to ensure that the behavior of readers works as expected. Clients of readers usually only tolerate `io.EOF` and not other errors -- so basically, I'm not yet sure if the expectation is that a closed reader should be considered "completed" or in an "error" state. I'm going to play around with it from the client perspective and if it makes sense to have a different error for this state then I will.

Reply:
If you are looking for existing idioms I would refer to net.ErrClosed or http.ErrServerClosed, both of which are used to indicate interruption of some operation by an extraneous Close-like call.

This is only a nit comment though, so don't waste too much energy on it.